adal/src/main/java/com/microsoft/aad/adal/BrokerAccountServiceHandler.java

```diff
@@ -324,13 +330,41 @@ private void bindToBrokerAccountService(final Context context, final Callback<Br
         if (brokerEvent != null) {
             brokerEvent.setBrokerAccountServiceBindingSucceed(serviceBound);
         }
+        if (!serviceBound) {
+            connection.unBindService(context);
+            callback.onError(new AuthenticationException(ADALError.BROKER_BIND_SERVICE_FAILED));
+        } else {
+            // make sure it's bound to the service with success in a period not longer than the TIMEOUT
+            mUiHandler.postDelayed(timeoutRunnable, BIND_SERVICE_TIMEOUT_THRESHOLD);
+        }
+    }
+
+    private class BindServiceTimeoutRunnable implements Runnable {
+        private Context mContext;
+        private BrokerAccountServiceConnection mServiceConnection;
+        private Callback<BrokerAccountServiceConnection> mCallback;
+
+        public BindServiceTimeoutRunnable(Context context, BrokerAccountServiceConnection connection, final Callback<BrokerAccountServiceConnection> callback) {
+            mContext = context;
+            mServiceConnection = connection;
+            mCallback = callback;
+        }
+
+        @Override
+        public void run() {
+            if (!mServiceConnection.isBound()) {
```

Comment:
It's unclear to me what's going on here. This appears to check that the service *isn't* bound before calling `unbindService(Context)`.

Reply:
the context is : there's some cases that the 'bindService' take too long, the callbacks of bindService ('onServiceConnected' ) were not called. So we added a timeout here. After 'bindService' was called, we set a timer of 30 seconds for now, if the service is bound with success in the 30 seconds, then cancel the timer in 'onServiceConnected'. If 'onServiceConnected' was not called in 30 seconds, then we unbind it in case it will block other requests since it's a single thread executor.